upgrades/steps_28.go

```diff
@@ -139,59 +156,163 @@ var getUpgradeStepsClient = func(caller base.APICaller) UpgradeStepsClient {
 
 //go:generate mockgen -package mocks -destination mocks/upgradestepsclient_mock.go github.com/juju/juju/upgrades UpgradeStepsClient
 type UpgradeStepsClient interface {
-	WriteUniterState(uniterStates map[names.Tag]string) error
+	WriteUniterState([]params.SetUnitStateArg) error
 }
 
 func iaasUniterState(ctx Context, unitNames []string) ([]string, error) {
 	// Read the uniter state for each unit on this machine.
 	// Leave in yaml format as a string to push to the controller.
-	fileNames := []string{}
-	uniterStates := make(map[names.Tag]string, len(unitNames))
+	fileNames := set.NewStrings()
+	uniterStates := make([]params.SetUnitStateArg, len(unitNames))
 	dataDir := ctx.AgentConfig().DataDir()
-	for _, unitName := range unitNames {
+	for i, unitName := range unitNames {
 		tag, err := names.ParseUnitTag(unitName)
 		if err != nil {
 			return nil, errors.Annotatef(err, "unable to parse unit agent tag %q", unitName)
 		}
 
+		uniterStates[i].Tag = tag.String()
 		// e.g. /var/lib/juju/agents/unit-ubuntu-0/state/uniter
-		uniterStateFile := filepath.Join(agent.BaseDir(dataDir), unitName, "state", "uniter")
+		uniterStateDir := filepath.Join(agent.BaseDir(dataDir), unitName, "state")
 
-		// Read into a State as we can validate the data.
-		var st operation.State
-		if err = utils.ReadYaml(uniterStateFile, &st); err != nil {
-			// No file available, move on.
-			if os.IsNotExist(err) {
-				continue
-			}
-			return nil, errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+		uniterSt, filename, err := readUniterState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
 		}
-		fileNames = append(fileNames, uniterStateFile)
-		if err = st.Validate(); err != nil {
-			return nil, errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+		if uniterSt != nil {
+			fileNames.Add(filename)
+			uniterStates[i].UniterState = uniterSt
 		}
 
-		data, err := yaml.Marshal(st)
-		if err != nil {
-			return nil, errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+		storageSt, filename, err := readStorageState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
+		}
+		if storageSt != nil {
+			fileNames.Add(filename)
+			uniterStates[i].StorageState = storageSt
 		}
-		uniterStates[tag] = string(data)
 	}
 
+	logger.Criticalf("filenames %d %s", fileNames.Size(), fileNames.Values())
 	// No state files, nothing to do.
-	if len(fileNames) == 0 {
+	if fileNames.IsEmpty() {
 		return nil, nil
 	}
 
 	client := getUpgradeStepsClient(ctx.APIState())
 	if err := client.WriteUniterState(uniterStates); err != nil {
 		return nil, errors.Annotatef(err, "unable to set state for units %q", strings.Join(unitNames, ", "))
 	}
-	return fileNames, nil
+	return fileNames.Values(), nil
+}
+
+func readUniterState(uniterStateDir string) (*string, string, error) {
```

Comment:
I would prefer that we did not return `*string` which looks odd to me (comment applies to returning pointers to scalar values in general).

How do you feel about changing `readUniterState` to return a small struct to encapsulate the state and statefile path and `errors.NotFound` to indicate missing state files?

Reply:
logic changed.